sys/vm/vm_mmap.c

```diff
@@ -1075,8 +1076,18 @@ kern_munmap(struct thread *td, uintptr_t addr0, size_t size)
 		}
 	}
 #endif
+	reservation = -1;
+	if (vm_map_lookup_entry(map, addr, &entry))
+		reservation = entry->reservation;
 	vm_map_delete(map, addr, addr + size);
 
+	result = vm_map_insert(map, NULL, 0, addr, addr + size,
+	    VM_PROT_NONE, VM_PROT_NONE, MAP_CREATE_UNMAPPED,
+	    reservation);
```

Comment:
Does this unhook pages from a vm object, say that's just got a single reference, when/if it splits a map entry, or does it leave the object in place and the pages effectively shadowed?  If the latter, could we call off to madvise's DONTNEED logic first?

Reply:
vm_map_delete rips the mappings out with pmap_remove call and then we insert a new UNMAPPED mapping in place.  I don't think there's a way to recover access to a parent's copy of a shadowed object the way there was with DONTNEED since there isn't a handle to the vm_object to remap in place.